elements/lisk-p2p/src/peer_directory/peer_book.ts

```diff
@@ -0,0 +1,136 @@
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+
+import { P2PDiscoveredPeerInfo, P2PPeerInfo } from '../p2p_types';
+import { NewPeerConfig, NewPeers } from './new_peers';
+import { TriedPeerConfig, TriedPeers } from './tried_peers';
+
+export interface PeerBookConfig {
+	readonly newPeerConfig?: NewPeerConfig;
+	readonly triedPeerConfig?: TriedPeerConfig;
+	readonly secret: number;
+}
+
+export class PeerBook {
+	private readonly _bannedPeers: ReadonlyArray<P2PDiscoveredPeerInfo>;
+	private readonly _newPeers: NewPeers;
+	private readonly _triedPeers: TriedPeers;
+	public constructor({
+		newPeerConfig,
+		triedPeerConfig,
+		secret,
+	}: PeerBookConfig) {
+		this._newPeers = new NewPeers(newPeerConfig ? newPeerConfig : { secret });
+		this._triedPeers = new TriedPeers(
+			triedPeerConfig ? triedPeerConfig : { secret },
+		);
+		this._bannedPeers = [];
+	}
+
+	public get newPeers(): ReadonlyArray<P2PPeerInfo> {
+		return this._newPeers.newPeersList();
+	}
+
+	public get triedPeers(): ReadonlyArray<P2PDiscoveredPeerInfo> {
+		return this._triedPeers.triedPeersList();
+	}
+
+	public get bannedPeers(): ReadonlyArray<P2PDiscoveredPeerInfo> {
+		return this._bannedPeers;
+	}
+
+	public getAllPeers(): ReadonlyArray<P2PPeerInfo> {
+		return [...this.newPeers, ...this.triedPeers];
+	}
+	/**
+	 * Description: When a peer is downgraded for some reasons then new/triedPeers will trigger their failedConnectionAction,
+	 * if the peer is deleted from newPeer that means the peer is completely deleted from the peer lists and need to inform the calling entity by returning true.
+	 */
+	public downgradePeer(peerInfo: P2PPeerInfo): boolean {
+		if (this._newPeers.getPeer(peerInfo)) {
+			if (this._newPeers.failedConnectionAction(peerInfo)) {
+				return true;
+			}
+		}
+
+		if (this._triedPeers.getPeer(peerInfo)) {
+			const failed = this._triedPeers.failedConnectionAction(peerInfo);
+			if (failed) {
+				this.addPeer(peerInfo);
+			}
+		}
+
+		return false;
+	}
+	// Move a peer from newPeer to triedPeer on events like on successful connection.
+	public upgradePeer(peerInfo: P2PPeerInfo): boolean {
+		if (this._triedPeers.getPeer(peerInfo)) {
+			return true;
+		}
+
+		if (this._newPeers.getPeer(peerInfo)) {
+			this._newPeers.removePeer(peerInfo);
+			this._triedPeers.addPeer(peerInfo as P2PDiscoveredPeerInfo);
+
+			return true;
+		}
+
+		return false;
+	}
+	// It will return evicted peer in the case a peer is removed from a peer list based on eviction strategy.
+	public addPeer(peerInfo: P2PPeerInfo): P2PPeerInfo | undefined {
+		if (
+			this._triedPeers.getPeer(peerInfo) ||
+			this._newPeers.getPeer(peerInfo)
+		) {
+			throw new Error('Peer already exists');
+		}
+
+		return this._newPeers.addPeer(peerInfo).evictedPeer;
```

Comment:
There are few peers like the fixed ones which we should add them straight to triedPeers. How do we do it?

Reply:
Whenever we have a fixed peer, we upgrade it first of all to be in the tried peer list once we start a node and we never downgrade it in any case